src/js/spatial-navigation.js

```diff
@@ -0,0 +1,496 @@
+
+      // Search inside array of 'focusableComponents' for a match of name of
+      // the last focused component.
+      for (let i = 0; i < this.focusableComponents.length; i++) {
+        if (this.focusableComponents[i].name_ === this.lastFocusedComponent.name_) {
+          this.focus(this.focusableComponents[i]);
+          return;
+        }
+      }
+    } else {
+      this.focus(this.getComponents()[0]);
+    }
+  }
+
+  /**
+   * Focuses on a given component.
+   * If the component is available to be focused, it focuses on the component.
+   * If not, it attempts to find a suitable DOM child within the component and focuses on it.
+   *
+   * @param {Component} component - The component to be focused.
+   */
+  focus(component) {
+    if (component.getIsAvailableToBeFocused(component.el_)) {
+      component.focus();
+    } else if (this.findSuitableDOMChild(component)) {
+      this.findSuitableDOMChild(component).focus();
+    }
+  }
+
+  /**
+   * Calculates the distance between two points, adjusting the calculation based on
+   * the specified navigation direction.
+   *
+   * @param {Object} center1 The center point of the first element.
+   * @param {Object} center2 The center point of the second element.
+   * @param {string} direction The direction of navigation ('up', 'down', 'left', 'right').
+   * @return {number} The calculated distance between the two centers.
+   */
+  calculateDistance(center1, center2, direction) {
+    const dx = Math.abs(center1.x - center2.x);
+    const dy = Math.abs(center1.y - center2.y);
+
+    let distance;
+
+    switch (direction) {
+    case 'right':
+    case 'left':
+      // Higher weight for vertical distance in horizontal navigation.
+      distance = dx + (dy * 100);
+      break;
+    case 'up':
+      // Strongly prioritize vertical proximity for UP navigation.
+      // Adjust the weight to ensure that elements directly above are favored.
+      distance = (dy * 2) + (dx * 0.5);
+      break;
+    case 'down':
+      // More balanced weight for vertical and horizontal distances.
+      // Adjust the weights here to find the best balance.
+      distance = (dy * 5) + dx;
+      break;
+    default:
+      distance = dx + dy;
+    }
+
+    return distance;
+  }
+
+  /**
+   * Adds an event listener for a specific event.
+   *
+   * @param {string} eventName - The name of the event to listen for.
+   * @param {Function} callback - The callback function to be executed when the event occurs.
+   */
+  addEventListener(eventName, callback) {
+    if (!this.eventListeners[eventName]) {
+      this.eventListeners[eventName] = [];
+    }
+
+    this.eventListeners[eventName].push(callback);
+  }
+
+  /**
+   * Notifies listeners for a specific event.
+   *
+   * @param {string} eventName - The name of the event to notify listeners for.
+   * @param {Object} [eventDetails={}] - Additional details to include in the event object.
+   */
+  notifyListeners(eventName, eventDetails = {}) {
```

Comment:
Could we take advantage EventTarget class and avoid some of this logic?

We could have the SpatialNavigation class extend `EventTarget`, and we could use the methods in that class to handle events

Reply:
Thanks for pointing this out @wseymour15 
I've changed SpatialNavigation class to extend EventTarget and removed the redundant methods.